In [1]:
# Import Library
import numpy as np
import pandas as pd

In [2]:
# Import the dataset
data = pd.read_csv("D:\DA\DA PROJECT\DA DATA\Project 4\dbo.customer.csv", encoding='utf-8')
data1 = pd.read_csv("D:\DA\DA PROJECT\DA DATA\Project 4\dim.date.csv", encoding='utf-8')
data2 = pd.read_csv("D:\DA\DA PROJECT\DA DATA\Project 4\dim.geography.csv", encoding='utf-8')
data3 = pd.read_csv("D:\DA\DA PROJECT\DA DATA\Project 4\dim.sales_territory.csv", encoding='utf-8')
data4 = pd.read_csv("D:\DA\DA PROJECT\DA DATA\Project 4\\fact_internet_sales.csv", encoding='utf-8')
data5 = pd.read_csv("D:\DA\DA PROJECT\DA DATA\Project 4\dim.product.csv", encoding='utf-8')

# Import dataset into list
data_list = [data, data1, data2, data3, data4, data5]

# Display the first 5 rows of the DataFrame in the list
for i, df in enumerate(data_list):
    print(f"DataFrame {i+1}:\n", df.head(), "\n")


DataFrame 1:
    CustomerKey  GeographyKey CustomerAlternateKey Title  FirstName MiddleName  \
0        11000            26           AW00011000   NaN        Jon          V   
1        11001            37           AW00011001   NaN     Eugene          L   
2        11002            31           AW00011002   NaN      Ruben        NaN   
3        11003            11           AW00011003   NaN    Christy        NaN   
4        11004            19           AW00011004   NaN  Elizabeth        NaN   

  LastName  NameStyle  BirthDate MaritalStatus  ... SpanishOccupation  \
0     Yang      False  10/6/1971             M  ...       Profesional   
1    Huang      False  5/10/1976             S  ...       Profesional   
2   Torres      False   2/9/1971             M  ...       Profesional   
3      Zhu      False  8/14/1973             S  ...       Profesional   
4  Johnson      False   8/5/1979             S  ...       Profesional   

  FrenchOccupation HouseOwnerFlag  NumberCarsOwned         A

In [3]:
# Eliminate unnecessary columns each dataframe in the list
for i, df in enumerate(data_list):
    if i == 0:  # dbo.customer
        data_list[i] = df[['CustomerKey', 'GeographyKey', 'FirstName', 'LastName', 'BirthDate', 'MaritalStatus', 'Gender', 'EmailAddress', 'Phone']]

    elif i == 1:  # dim.date
        #  No columns to eliminate
        data_list[i] = data_list[i] 
    elif i == 2:  # dim.geography
        data_list[i] = df[['GeographyKey', 'City', 'EnglishCountryRegionName']]
    elif i == 3:  # dim.sales_territory
        data_list[i] = data_list[i]
    elif i == 4:  # fact_internet_sales
        data_list[i] = data_list[i]
    elif i == 5:  # dim.product
        data_list[i] = df[['ProductKey', 'EnglishProductName']]

In [4]:
#Check for missing values in each DataFrame
for i, df in enumerate(data_list):
    missing_values = df.isnull().sum()
    print(f"Missing values in DataFrame {i+1}:\n{missing_values}\n")

Missing values in DataFrame 1:
CustomerKey         0
GeographyKey        0
FirstName           0
LastName            0
BirthDate           0
MaritalStatus       0
Gender              0
EmailAddress        0
Phone            1772
dtype: int64

Missing values in DataFrame 2:
DateKey                 0
FullDateAlternateKey    0
DayNumberOfWeek         0
EnglishDayNameOfWeek    0
SpanishDayNameOfWeek    0
FrenchDayNameOfWeek     0
DayNumberOfMonth        0
DayNumberOfYear         0
WeekNumberOfYear        0
EnglishMonthName        0
SpanishMonthName        0
FrenchMonthName         0
MonthNumberOfYear       0
CalendarQuarter         0
CalendarYear            0
CalendarSemester        0
FiscalQuarter           0
FiscalYear              0
FiscalSemester          0
dtype: int64

Missing values in DataFrame 3:
GeographyKey                0
City                        0
EnglishCountryRegionName    0
dtype: int64

Missing values in DataFrame 4:
SalesTerritoryKey              0
SalesTerritoryAlter

<H3> DataFrame 1: Column Phone is missing 1772 values, but it doesnt neccessary to fix. <br> DataFrame 4: Column SalesTerritoryRegion, SalesTerritoryCoutry, SalesTerritoryGroup missing one, but it doesnt neccessary to fix. <br> DataFrame 5: CarrierTrackingNumber and CustomerPONumber missing, but it doesnt neccessary to fix. <H3>

In [5]:
# Check the data types of each DataFrame
for i, df in enumerate(data_list):
    print(f"Data types in DataFrame {i+1}:\n {df.dtypes}\n")

Data types in DataFrame 1:
 CustomerKey       int64
GeographyKey      int64
FirstName        object
LastName         object
BirthDate        object
MaritalStatus    object
Gender           object
EmailAddress     object
Phone            object
dtype: object

Data types in DataFrame 2:
 DateKey                  int64
FullDateAlternateKey    object
DayNumberOfWeek          int64
EnglishDayNameOfWeek    object
SpanishDayNameOfWeek    object
FrenchDayNameOfWeek     object
DayNumberOfMonth         int64
DayNumberOfYear          int64
WeekNumberOfYear         int64
EnglishMonthName        object
SpanishMonthName        object
FrenchMonthName         object
MonthNumberOfYear        int64
CalendarQuarter          int64
CalendarYear             int64
CalendarSemester         int64
FiscalQuarter            int64
FiscalYear               int64
FiscalSemester           int64
dtype: object

Data types in DataFrame 3:
 GeographyKey                 int64
City                        object
EnglishCoun

In [6]:
# Change the data types of specific columns
data_list[1]["FullDateAlternateKey"] = pd.to_datetime(data_list[1]["FullDateAlternateKey"], format="%m/%d/%Y")

In [7]:
# Check again the data types after conversion
print("Data types after conversion in DataFrame 2:\n", data_list[1].dtypes)

Data types after conversion in DataFrame 2:
 DateKey                          int64
FullDateAlternateKey    datetime64[ns]
DayNumberOfWeek                  int64
EnglishDayNameOfWeek            object
SpanishDayNameOfWeek            object
FrenchDayNameOfWeek             object
DayNumberOfMonth                 int64
DayNumberOfYear                  int64
WeekNumberOfYear                 int64
EnglishMonthName                object
SpanishMonthName                object
FrenchMonthName                 object
MonthNumberOfYear                int64
CalendarQuarter                  int64
CalendarYear                     int64
CalendarSemester                 int64
FiscalQuarter                    int64
FiscalYear                       int64
FiscalSemester                   int64
dtype: object


<H3> Data type in each DataFrame now are appropriate <H3>

In [8]:
# Check for duplicate rows in each DataFrame
for i, df in enumerate(data_list):
    duplicate_rows = df.duplicated().sum()
    print(f"Number of duplicate rows in DataFrame {i+1}: {duplicate_rows}\n")

Number of duplicate rows in DataFrame 1: 0

Number of duplicate rows in DataFrame 2: 0

Number of duplicate rows in DataFrame 3: 0

Number of duplicate rows in DataFrame 4: 0

Number of duplicate rows in DataFrame 5: 0

Number of duplicate rows in DataFrame 6: 0



In [9]:
# Delete extra spaces in string columns in each DataFrame
for i, df in enumerate(data_list):
    df = df.apply(lambda x: x.str.strip() if x.dtype == "object" else x)
    data_list[i] = df # Update the DataFrame in the list
# Display the cleaned DataFrames
for i, df in enumerate(data_list):
    print(f"Cleaned DataFrame {i+1}:\n", df.head(), "\n")

Cleaned DataFrame 1:
    CustomerKey  GeographyKey  FirstName LastName  BirthDate MaritalStatus  \
0        11000            26        Jon     Yang  10/6/1971             M   
1        11001            37     Eugene    Huang  5/10/1976             S   
2        11002            31      Ruben   Torres   2/9/1971             M   
3        11003            11    Christy      Zhu  8/14/1973             S   
4        11004            19  Elizabeth  Johnson   8/5/1979             S   

  Gender                    EmailAddress                Phone  
0      M       jon24@adventure-works.com  1 (11) 500 555-0162  
1      M    eugene10@adventure-works.com  1 (11) 500 555-0110  
2      M     ruben35@adventure-works.com  1 (11) 500 555-0184  
3      F   christy12@adventure-works.com  1 (11) 500 555-0162  
4      F  elizabeth5@adventure-works.com  1 (11) 500 555-0131   

Cleaned DataFrame 2:
     DateKey FullDateAlternateKey  DayNumberOfWeek EnglishDayNameOfWeek  \
0  20050101           2005-01-01 

In [10]:
# Exprort the cleaned DataFrames to CSV files
for i, df in enumerate(data_list):
    if i == 0:
        df.to_csv(f"D:/DA/DA PROJECT/DA DATA/Project 4/cleaned_customer.csv", index=False, encoding='utf-8')
    if i == 1:
        df.to_csv(f"D:/DA/DA PROJECT/DA DATA/Project 4/cleaned_date.csv", index=False, encoding='utf-8')
    if i == 2:
        df.to_csv(f"D:/DA/DA PROJECT/DA DATA/Project 4/cleaned_geography.csv", index=False, encoding='utf-8')
    if i == 3:
        df.to_csv(f"D:/DA/DA PROJECT/DA DATA/Project 4/cleaned_sales_territory.csv", index=False, encoding='utf-8')
    if i == 4:
        df.to_csv(f"D:/DA/DA PROJECT/DA DATA/Project 4/cleaned_fact_internet_sales.csv", index=False, encoding='utf-8')
    if i == 5:
        df.to_csv(f"D:/DA/DA PROJECT/DA DATA/Project 4/cleaned_product.csv", index=False, encoding='utf-8')